lib/utils.ts

```diff
@@ -395,6 +419,43 @@ async function requestAsync(
 		opts.headers = nativeHeaders;
 	}
 
+	const form = getForm();
+	let hasBlob = false;
+	for (const [k, v] of Object.entries<string | WebResourceFile>(
+		options.body ?? {},
+	)) {
+		if (isFile(v)) {
+			hasBlob = true;
+			form.append(k, v, v.name);
+		} else {
+			form.append(k, v);
+		}
+	}
+
+	if (hasBlob) {
+		if (IS_BROWSER) {
+			// Browsers will handle set form data header and boundaries
+			// Given the correct body format
+			opts.headers.delete('Content-Type');
+
+			opts.body = form as FormData;
+		} else {
+			const { FormDataEncoder } =
+				// eslint-disable-next-line @typescript-eslint/no-var-requires
+				require('form-data-encoder') as typeof import('form-data-encoder');
+
+			const encoder = new FormDataEncoder(form as FormDataNodeType);
+			opts.headers.set('Content-Type', encoder.headers['Content-Type']);
+			const length = encoder.headers['Content-Length'];
+			if (length != null) {
+				opts.headers.set('Content-Length', length);
+			}
+
+			// @ts-expect-error https://www.npmjs.com/package/form-data-encoder#usage
+			opts.body = Readable.from(encoder);
+		}
+	}
```

Comment:
How about v so that we do not create & populate the form object unnecessarily?
```suggestion
	const bodyEntries = Object.entries<string | WebResourceFile>(options.body ?? {});
	const fileKeys = new Set<string>();

	for (const [k, v] of bodyEntries) {
		if (isFile(v)) {
			fileKeys.add(k);
		}
	}

	if (fileKeys.length > 0) {
		const form = getForm();
		for (const [k, v] of bodyEntries) {
			if (fileKeys.has(k)) {
				form.append(k, v, v.name);
			} else {
				form.append(k, v);
			}
		}

		if (IS_BROWSER) {
			// Browsers will handle set form data header and boundaries
			// Given the correct body format
			opts.headers.delete('Content-Type');

			opts.body = form as FormData;
		} else {
			const encoder = new FormDataEncoder(form as typeof nodeForm);
			opts.headers.set('Content-Type', encoder.headers['Content-Type']);
			const length = encoder.headers['Content-Length'];
			if (length != null) {
				opts.headers.set('Content-Length', length);
			}

			// @ts-expect-error https://www.npmjs.com/package/form-data-encoder#usage
			opts.body = Readable.from(encoder);
		}
	}
```

Reply:
@thgreasi  Well, I don't see then the (complexity) difference from that and my initial approach which was similar to this one See: https://github.com/balena-io-modules/balena-request/pull/175#discussion_r1346929752

Reply:
@otaviojacobi I back then I hadn't realized that ALL properties have to end up in the multipart form when there is at least one file. I'm sorry for the confusion :(

Reply:
Np! thanks!